File: Data Structures & Algorithms/trapping-rain-water/submission-1.py
class Solution:
    def trap(self, height: List[int]) -> int:
        curr_max=height[0]
        l=len(height)
        pref_max,suff_max=[],[]
        res=0
        for i in range(len(height)):
            if height[i]>curr_max:
                curr_max=height[i]
            pref_max.append(curr_max)
        curr_max=height[l-1]
        for i in range(len(height)-1,-1,-1):
            if height[i]>curr_max:
                curr_max=height[i]
            suff_max.append(curr_max)
        suff_max=suff_max[::-1]
        for i in range(1,l-1):
            res+=max(min(pref_max[i], suff_max[i]) - height[i],0)
        return res

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-1.cpp
class Solution {
    vector<int> dp;
public:
    // int f(int n,vector<int>& cost){
    //     if(dp[n]!=-1) return dp[n];
    //     int l=f(n-1,cost);
    //     int r=f(n-2,cost);
    //     return dp[n] = min(l+cost[n-1],r+cost[n-2]);

    // }
    int minCostClimbingStairs(vector<int>& cost) {
        int n=cost.size();
        dp=vector<int>(n+1,-1);
        dp[0]=0;dp[1]=0;
        for(int i=2;i<=n;i++){
            dp[i]=min(dp[i-1]+cost[i-1],dp[i-2]+cost[i-2]);
        }
        return dp[n];


        
    }
};
